File: RxHttp/src/main/java/com/github/kongpf8848/rxhttp/RxHttp.java
package com.github.kongpf8848.rxhttp;

import android.app.Activity;
import android.content.Context;
import android.text.TextUtils;

import androidx.fragment.app.Fragment;
import androidx.lifecycle.Lifecycle;
import androidx.lifecycle.LifecycleOwner;

import com.github.kongpf8848.rxhttp.bean.DownloadInfo;
import com.github.kongpf8848.rxhttp.callback.DownloadCallback;
import com.github.kongpf8848.rxhttp.callback.HttpCallback;
import com.github.kongpf8848.rxhttp.converter.DownloadConverter;
import com.github.kongpf8848.rxhttp.converter.GsonConverter;
import com.github.kongpf8848.rxhttp.request.AbsRequest;
import com.github.kongpf8848.rxhttp.request.DeleteRequest;
import com.github.kongpf8848.rxhttp.request.DownloadRequest;
import com.github.kongpf8848.rxhttp.request.GetRequest;
import com.github.kongpf8848.rxhttp.request.PostFormRequest;
import com.github.kongpf8848.rxhttp.request.PostRequest;
import com.github.kongpf8848.rxhttp.request.PutRequest;
import com.github.kongpf8848.rxhttp.request.UploadRequest;
import com.github.kongpf8848.rxhttp.util.Md5Util;
import com.trello.rxlifecycle2.LifecycleTransformer;
import com.uber.autodispose.AutoDispose;
import com.uber.autodispose.android.lifecycle.AndroidLifecycleScopeProvider;

import java.io.File;

import io.reactivex.Observable;
import io.reactivex.ObservableSource;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.functions.Function;
import io.reactivex.schedulers.Schedulers;
import okhttp3.OkHttpClient;
import okhttp3.ResponseBody;
import retrofit2.Retrofit;
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory;
import retrofit2.converter.gson.GsonConverterFactory;
import retrofit2.converter.scalars.ScalarsConverterFactory;

public class RxHttp {

    private OkHttpClient okHttpClient;
    private Retrofit retrofit;
    private HttpInterceptor interceptor;
    private HttpService httpService;
    private static RxHttp instance;

    public static RxHttp getInstance() {
        if (instance == null) {
            synchronized (RxHttp.class) {
                if (instance == null) {
                    instance = new RxHttp();
                }
            }
        }
        return instance;
    }


    private RxHttp() {
        RxHttpConfig config = RxHttpConfig.getInstance();


        OkHttpClient.Builder builder = config.getBuilder();
        interceptor = new HttpInterceptor();
        builder.addInterceptor(interceptor);
        okHttpClient = builder.build();
        retrofit = new Retrofit.Builder()
                .client(okHttpClient)
                .addConverterFactory(ScalarsConverterFactory.create())
                .addConverterFactory(GsonConverterFactory.create())
                .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
                .baseUrl("http://example.com")
                .build();
        httpService = retrofit.create(HttpService.class);
    }


    //get请求
    public GetRequest get(Context context) {
        return new GetRequest(context);
    }

    public GetRequest get(Activity activity) {
        return new GetRequest(activity);
    }

    public GetRequest get(Fragment fragment) {
        return new GetRequest(fragment);
    }

    public GetRequest get(LifecycleTransformer transformer) {
        return new GetRequest(transformer);
    }

    //post请求
    public PostRequest post(Context context) {
        return new PostRequest(context);
    }

    public PostRequest post(Activity activity) {
        return new PostRequest(activity);
    }

    public PostRequest post(Fragment fragment) {
        return new PostRequest(fragment);
    }

    public PostRequest post(LifecycleTransformer transformer) {
        return new PostRequest(transformer);
    }

    //post表单请求
    public PostFormRequest postForm(Context context) {
        return new PostFormRequest(context);
    }

    public PostFormRequest postForm(Activity activity) {
        return new PostFormRequest(activity);
    }

    public PostFormRequest postForm(Fragment fragment) {
        return new PostFormRequest(fragment);
    }

    public PostFormRequest postForm(LifecycleTransformer transformer) {
        return new PostFormRequest(transformer);
    }

    //upload请求
    public UploadRequest upload(Context context) {
        return new UploadRequest(context);
    }

    public UploadRequest upload(Activity activity) {
        return new UploadRequest(activity);
    }

    public UploadRequest upload(Fragment fragment) {
        return new UploadRequest(fragment);
    }

    public UploadRequest upload(LifecycleTransformer transformer) {
        return new UploadRequest(transformer);
    }

    //download请求
    public DownloadRequest download(Context context) {
        return new DownloadRequest(context);
    }

    public DownloadRequest download(Activity activity) {
        return new DownloadRequest(activity);
    }

    public DownloadRequest download(Fragment fragment) {
        return new DownloadRequest(fragment);
    }

    public DownloadRequest download(LifecycleTransformer transformer) {
        return new DownloadRequest(transformer);
    }

    //put请求
    public PutRequest put(Context context) {
        return new PutRequest(context);
    }
    public PutRequest put(Activity activity) {
        return new PutRequest(activity);
    }
    public PutRequest put(Fragment fragment) {
        return new PutRequest(fragment);
    }
    public PutRequest put(LifecycleTransformer transformer) {
        return new PutRequest(transformer);
    }

    //delete请求
    public DeleteRequest delete(Context context) {
        return new DeleteRequest(context);
    }
    public DeleteRequest delete(Activity activity) {
        return new DeleteRequest(activity);
    }
    public DeleteRequest delete(Fragment fragment) {
        return new DeleteRequest(fragment);
    }
    public DeleteRequest delete(LifecycleTransformer transformer) {
        return new DeleteRequest(transformer);
    }

    public <T> void enqueue(final AbsRequest request, final HttpCallback<T> callback) {
        interceptor.setRequest(request);
        Observable<ResponseBody> observable = null;
        if (request instanceof GetRequest) {
            observable = httpService.get(request.getUrl(),request.getParams());
        }
        else if(request instanceof PutRequest){
            PutRequest putRequest = (PutRequest) request;
            observable = httpService.put(putRequest.getUrl(), putRequest.buildRequestBody());
        }
        else if (request instanceof PostRequest) {
            PostRequest postRequest = (PostRequest) request;
            observable = httpService.post(postRequest.getUrl(), postRequest.buildRequestBody());
        }
        else if (request instanceof PostFormRequest) {
            observable = httpService.postForm(request.getUrl(), request.getParams());
        }
        else if(request instanceof DeleteRequest){
            observable = httpService.delete(request.getUrl(), request.getParams());
        }
        else if (request instanceof UploadRequest) {
            final UploadRequest uploadRequest = (UploadRequest) request;
            observable = httpService.post(uploadRequest.getUrl(), uploadRequest.buildRequestBody());
        } else if (request instanceof DownloadRequest) {
            final DownloadRequest downloadRequest = ((DownloadRequest) request);
            final File file=new File(downloadRequest.getDir(),downloadRequest.getFilename());
            DownloadInfo downloadInfo=new DownloadInfo(downloadRequest.getUrl(),downloadRequest.getDir(),downloadRequest.getFilename());
            if(!TextUtils.isEmpty(downloadRequest.getMd5())){
                if(file.exists()){
                    String fileMd5= Md5Util.getMD5(file);
                    if(downloadRequest.getMd5().equalsIgnoreCase(fileMd5)){
                        downloadInfo.setTotal(file.length());
                        downloadInfo.setProgress(file.length());
                        callback.onNext((T)downloadInfo);
                        return;
                    }
                }
            }
            if (downloadRequest.isBreakpoint()) {
                observable=Observable.just(downloadRequest.getUrl())
                        .flatMap(new Function<String, ObservableSource<Long>>() {
                    @Override
                    public ObservableSource<Long> apply(String url) throws Exception {
                        long contentLength = getContentLength(url);
                        long start=0;
                        if(file.exists()){
                            if(contentLength==-1 || file.length()>=contentLength){
                                file.delete();
                            }
                            else{
                                start=file.length();
                            }
                        }
                        return Observable.just(start);
                    }
                }).flatMap(new Function<Long, ObservableSource<ResponseBody>>() {
                    @Override
                    public ObservableSource<ResponseBody> apply(Long start) throws Exception {
                        return httpService.download(downloadRequest.getUrl(),String.format("bytes=%d-",start));
                    }
                }).subscribeOn(Schedulers.io());

            } else {
                observable = httpService.download(request.getUrl());
            }
        }


        if (observable != null) {
            Observable<T> observableFinal = observable.map(new Function<ResponseBody, T>() {
                @Override
                public T apply(ResponseBody body) throws Exception {
                    if (request instanceof DownloadRequest) {
                        DownloadConverter<T> downloadConverter = new DownloadConverter((DownloadRequest) request, (DownloadCallback) callback);
                        return downloadConverter.convert(body, callback.getType());
                    } else {
                        return new GsonConverter<T>().convert(body, callback.getType());

                    }
                }
            }).retryWhen(new RetryWithDelay(RxHttpConfig.getInstance().getMaxRetries(), RxHttpConfig.getInstance().getRetryDelayMillis()))
                    .subscribeOn(Schedulers.io())
                    .observeOn(AndroidSchedulers.mainThread());
            if (request.getContext() instanceof LifecycleOwner) {
                LifecycleOwner lifecycleOwner = (LifecycleOwner) request.getContext();
                observableFinal.as(AutoDispose.<T>autoDisposable(AndroidLifecycleScopeProvider.from(lifecycleOwner, Lifecycle.Event.ON_DESTROY)))
                        .subscribeWith(new HttpObserver<T>(callback));
            } else if (request.getContext() instanceof LifecycleTransformer) {
                observableFinal.compose((LifecycleTransformer) request.getContext())
                        .subscribeWith(new HttpObserver<T>(callback));
            } else {
                observableFinal.subscribeWith(new HttpObserver<T>(callback));
            }


        }

    }


    private long getContentLength(String url) throws Exception {
        retrofit2.Response<Void> response=httpService.head(url).execute();
        if(response.isSuccessful()){
            String contentLength=response.headers().get("Content-Length");
            if(!TextUtils.isEmpty(contentLength)){
                return Long.parseLong(contentLength);
            }
        }
        return -1;
//        Request request = new Request.Builder()
//                .url(url)
//                .build();
//        Response response = new OkHttpClient.Builder().build().newCall(request).execute();
//        if (response.isSuccessful()) {
//            long contentLength = response.body().contentLength();
//            response.close();
//            return contentLength;
//        }
//        return -1;
    }


}

File: RxHttp/src/main/java/com/github/kongpf8848/rxhttp/RxHttp.kt
package com.github.kongpf8848.rxhttp

import android.app.Activity
import android.content.Context
import android.text.TextUtils
import androidx.fragment.app.Fragment
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.LifecycleOwner
import com.github.kongpf8848.rxhttp.bean.DownloadInfo
import com.github.kongpf8848.rxhttp.callback.DownloadCallback
import com.github.kongpf8848.rxhttp.callback.HttpCallback
import com.github.kongpf8848.rxhttp.converter.DownloadConverter
import com.github.kongpf8848.rxhttp.converter.GsonConverter
import com.github.kongpf8848.rxhttp.request.*
import com.github.kongpf8848.rxhttp.util.Md5Util
import com.uber.autodispose.AutoDispose
import com.uber.autodispose.android.lifecycle.AndroidLifecycleScopeProvider
import io.reactivex.Observable
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.functions.Function
import io.reactivex.schedulers.Schedulers
import okhttp3.OkHttpClient
import okhttp3.ResponseBody
import retrofit2.Retrofit
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory
import retrofit2.converter.gson.GsonConverterFactory
import retrofit2.converter.scalars.ScalarsConverterFactory
import java.io.File

class RxHttp private constructor() {

    private object RxHttp {
        val holder = RxHttp()
    }

    private val okHttpClient: OkHttpClient
    private val retrofit: Retrofit
    private val httpService: HttpService

    init {
        val config = RxHttpConfig.getInstance()
        val builder = config.getBuilder()
        okHttpClient = builder.build()
        retrofit = Retrofit.Builder()
                .client(okHttpClient)
                .addConverterFactory(ScalarsConverterFactory.create())
                .addConverterFactory(GsonConverterFactory.create())
                .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
                .baseUrl("http://example.com")
                .build()
        httpService = retrofit.create(HttpService::class.java)
    }

    //GET请求
    fun get(context: Context): GetRequest {
        return GetRequest(context)
    }

    fun get(activity: Activity): GetRequest {
        return GetRequest(activity)
    }

    fun get(fragment: Fragment): GetRequest {
        return GetRequest(fragment)
    }

    //POST请求
    fun post(context: Context): PostRequest {
        return PostRequest(context)
    }

    fun post(activity: Activity): PostRequest {
        return PostRequest(activity)
    }

    fun post(fragment: Fragment): PostRequest {
        return PostRequest(fragment)
    }

    //POST FORM请求
    fun postForm(context: Context): PostFormRequest {
        return PostFormRequest(context)
    }

    fun postForm(activity: Activity): PostFormRequest {
        return PostFormRequest(activity)
    }

    fun postForm(fragment: Fragment): PostFormRequest {
        return PostFormRequest(fragment)
    }

    //Upload请求
    fun upload(context: Context): UploadRequest {
        return UploadRequest(context)
    }

    fun upload(activity: Activity): UploadRequest {
        return UploadRequest(activity)
    }

    fun upload(fragment: Fragment): UploadRequest {
        return UploadRequest(fragment)
    }

    //Download请求
    fun download(context: Context): DownloadRequest {
        return DownloadRequest(context)
    }

    fun download(activity: Activity): DownloadRequest {
        return DownloadRequest(activity)
    }

    fun download(fragment: Fragment): DownloadRequest {
        return DownloadRequest(fragment)
    }

    //PUT请求
    fun put(context: Context): PutRequest {
        return PutRequest(context)
    }

    fun put(activity: Activity): PutRequest {
        return PutRequest(activity)
    }

    fun put(fragment: Fragment): PutRequest {
        return PutRequest(fragment)
    }

    //DELETE请求
    fun delete(context: Context): DeleteRequest {
        return DeleteRequest(context)
    }

    fun delete(activity: Activity): DeleteRequest {
        return DeleteRequest(activity)
    }

    fun delete(fragment: Fragment): DeleteRequest {
        return DeleteRequest(fragment)
    }

    fun <T> enqueue(request: AbsRequest, callback: HttpCallback<T>) {
        var observable: Observable<ResponseBody>? = null
        if (request is GetRequest) {
            observable = if (request.getParams() != null) {
                httpService.get(request.url, request.getParams()!!)
            } else {
                httpService.get(request.url)
            }
        } else if (request is PutRequest) {
            observable = httpService.put(request.url, request.buildRequestBody()!!)
        } else if (request is UploadRequest) {
            observable = httpService.post(request.url, request.buildRequestBody()!!)
        } else if (request is PostRequest) {
            observable = httpService.post(request.url, request.buildRequestBody()!!)
        } else if (request is PostFormRequest) {
            observable = if (request.getParams() != null) {
                httpService.postForm(request.url, request.getParams()!!)
            } else {
                httpService.postForm(request.url)
            }
        } else if (request is DeleteRequest) {
            observable = if (request.getParams() != null) {
                httpService.delete(request.url, request.getParams()!!)
            } else {
                httpService.delete(request.url)
            }
        } else if (request is DownloadRequest) {
            val file = File(request.dir, request.filename)
            val downloadInfo = DownloadInfo(request.url, request.dir, request.filename)
            if (!TextUtils.isEmpty(request.md5)) {
                if (file.exists()) {
                    val fileMd5 = Md5Util.getMD5(file)
                    if (request.md5.equals(fileMd5, ignoreCase = true)) {
                        downloadInfo.total = file.length()
                        downloadInfo.progress = file.length()
                        callback.onNext(downloadInfo as T)
                        return
                    }
                }
            }
            if (request.isBreakpoint) {
                observable = Observable.just(request.url)
                        .flatMap { url ->
                            val contentLength = getContentLength(url)
                            var startPosition: Long = 0
                            if (file.exists()) {
                                if (contentLength == -1L || file.length() >= contentLength) {
                                    file.delete()
                                } else {
                                    startPosition = file.length()
                                }
                            }
                            Observable.just(startPosition)
                        }.flatMap { position -> httpService.download(request.url, String.format("bytes=%d-", position)) }.subscribeOn(Schedulers.io())
            } else {
                observable = httpService.download(request.url)
            }
        }
        if (observable != null) {
            val observableFinal = observable.map(Function<ResponseBody, T> { body ->
                if (request is DownloadRequest) {
                    val downloadConverter: DownloadConverter<T> = DownloadConverter(request, callback as DownloadCallback)
                    downloadConverter.convert(body, callback.type)
                } else {
                    GsonConverter<T>().convert(body, callback.type)
                }
            }).retryWhen(RetryWithDelay(RxHttpConfig.getInstance().maxRetries, RxHttpConfig.getInstance().retryDelayMillis))
                    .subscribeOn(Schedulers.io())
                    .observeOn(AndroidSchedulers.mainThread())
            if (request.context is LifecycleOwner) {
                val lifecycleOwner = request.context as LifecycleOwner
                observableFinal.`as`(AutoDispose.autoDisposable(AndroidLifecycleScopeProvider.from(lifecycleOwner, Lifecycle.Event.ON_DESTROY)))
                        .subscribe(HttpObserver(callback, request.getTag()))
            } else {
                observableFinal.subscribeWith(HttpObserver(callback, request.getTag()))
            }
        }
    }

    @Throws(Exception::class)
    private fun getContentLength(url: String): Long {
        val response = httpService.head(url).execute()
        if (response.isSuccessful) {
            val contentLength = response.headers()["Content-Length"]
            if (!TextUtils.isEmpty(contentLength)) {
                return contentLength!!.toLong()
            }
        }
        return -1
    }

    /**
     * 取消网络请求,tag为null取消所有网络请求,tag不为null值取消指定的网络请求
     */
    fun cancelRequest(tag: Any?) {
        RxHttpTagManager.getInstance().cancelTag(tag)
    }

    companion object {
        fun getInstance() = RxHttp.holder
    }

}